### PR TITLE
ENH: Sliding_P barostat for ASE by Alexander Held (L. Pastewka et al., Tribol. Lett. 39, 49 (2010))

### PR DESCRIPTION
**Pull request by threich**
ASE version of the Sliding_P barostat by Lars Pastewka as implemented by Alexander Held.

### CODE REVIEW

File: examples/dftbpsocket/dftb_in.hsd [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    0. 0. 0.
+    25.29166 0. 0.
+    0. 25.29166 0.
+    0. 0. 25.29166
+}
+
+# Socket controlled DFTB+
+Driver = Socket {
+    File = "zundel" # File name, i.e., /tmp/ipi_dftb-zundel
+    MaxSteps = -1 # keep running until i-PI terminates the run
```
Comment (pastewka) [inline examples/dftbpsocket/dftb_in.hsd]: Please remove reference to i-PI. What is the i-PI license, can we just reuse something here that is taken from them?

File: matscipy/dftbp_socketcalc.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+# ======================================================================
+
+"""The dftbsocket module.
+
+This python module provides an ASE calculator for DFTB+.
+It operates DFTB+ through the i-PI interface protocol, which is implemented in DFTB+.
+i-PI can be obtained from https://github.com/i-pi/i-pi/tree/master.
+The i-PI dependency is not (yet) compatible with python 3,
+so the DFTB+ wrapper will not work with python 3 until i-PI is fixed.
```
Comment (pastewka) [inline matscipy/dftbp_socketcalc.py]: Why is this the case? Even if i-PI is not Python 3, the socket calculator part is probably Fortran? I guess there is a patch for DFTB+ at that location. Please update documentation that this becomes clear.

File: matscipy/dftbp_socketcalc.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+A usage example is found in the example directory.
+Please mind that this implementation is NOT WELL TESTED!!!
```
Comment (pastewka) [inline matscipy/dftbp_socketcalc.py]: Nothing is well tested, please remove.

File: matscipy/dftbp_socketcalc.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+Make sure to use the same socket specification in the ASE calculator and in the dftb+ input file!
+Unix sockets and network sockets are supported (only unix sockets tested so far).
+Please also read the Socket Driver section in the DFTB+ user manual.
```
Comment (pastewka) [inline matscipy/dftbp_socketcalc.py]: Okay...

File: matscipy/dftbp_socketcalc.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+"""
+
+from __future__ import (
+    division,
+    absolute_import,
+    print_function,
+    unicode_literals
+)
+from ase.calculators.calculator import Calculator, all_changes
+from ipi.engine.forcefields import FFSocket
```
Comment (pastewka) [inline matscipy/dftbp_socketcalc.py]: I see, this actually uses iPI. Why is this necessary? I think it is not a good idea to have a dependency on a third part package such as iPI in here.

@jameskermode What is your opinion? I think this should be properly ported to be matscipy only.

File: tests/test_dftbplus.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
+#! /usr/bin/env python
+
+# ======================================================================
+# matscipy - Python materials science tools
+# https://github.com/libAtoms/matscipy
+#
+# Copyright (2014) James Kermode, King's College London
```
Comment (pastewka) [inline tests/test_dftbplus.py]: Note that this copyright notice should reflect whoever wrote the present code.

**Comment by pastewka** [comment]
This looks like it is not the SlidingP barostat, but the DFTB+ socket calculator. Please update the title of the merge request accordingly.

**Comment by pastewka** [comment]
PS: I see that it also contains the barostat. @threich Can you split this into two merge request. I don't think the socket calculator is ready for merge, the barostat may be.

File: examples/sliding_p/equilibrate_pressure.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+# if we start from local minimum, zero potential energy, use double temperature for
+# faster temperature convergence in the beginning:
+MaxwellBoltzmannDistribution(atoms, 2 * kB * T)
+# clear momenta in constraint regions, otherwise lid might run away
+atoms.arrays['momenta'][top_mask, :] = 0
+atoms.arrays['momenta'][bottom_mask, :] = 0
+
+calc = ASE_CALCULATOR_OBJECT  # put a specific calculator here
+
+atoms.calc = calc
```
Comment (pastewka) [inline examples/sliding_p/equilibrate_pressure.py]: Shouldn't this be `set_calculator`?

File: matscipy/sliding_p.py [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+    """Automatic damping
+
+    Following L. Pastewka, S. Moser, and M. Moseler,
+    Tribol. Lett. 39, 49 (2010).
+    """
+
+    def __init__(self, C11, p_c=0.01):
+        """Constructor.
+
+        Arguments as described in L. Pastewka, S. Moser, and M. Moseler,
```
Comment (pastewka) [inline matscipy/sliding_p.py]: Please be explicit here what the arguments are.

File: matscipy/sliding_p.py [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+        Tribol. Lett. 39, 49 (2010).
+        """
+        self.C11 = float(C11)
+        self.p_c = float(p_c)
+
+    def get_M_gamma(self, slider, atoms):
```
Comment (pastewka) [inline matscipy/sliding_p.py]: documentation missing

File: matscipy/sliding_p.py [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+        M = k * omega_c ** -2 * np.sqrt(self.p_c ** -2 - 1.)
+        # gamma = np.sqrt(2. * M * k)  # <- expression from paper, but this is wrong
+        gamma = 2. * np.sqrt(M * k)
+        return M, gamma
+
+
+class FixedDamping(object):
+    """Damping with fixed damping constant and fixed mass."""
+
+    def __init__(self, gamma, M_factor=1.0):
```
Comment (pastewka) [inline matscipy/sliding_p.py]: documentation missing

File: matscipy/sliding_p.py [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+        self.gamma = float(gamma)
+        self.M_factor = float(M_factor)
+
+    def get_M_gamma(self, slider, atoms):
```
Comment (pastewka) [inline matscipy/sliding_p.py]: documentation missing

File: matscipy/sliding_p.py [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+        return M_top * self.M_factor, self.gamma
+
+
+class FixedMassCriticalDamping(object):
+    """Damping with fixed mass and critical damping constant.
+
+    Useful for fast pressure equilibration with small lid mass.
+    """
+
+    def __init__(self, C11, M_factor=1.0):
```
Comment (pastewka) [inline matscipy/sliding_p.py]: documentation missing

File: matscipy/sliding_p.py [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+        self.C11 = float(C11)
+        self.M_factor = float(M_factor)
+
+    def get_M_gamma(self, slider, atoms):
```
Comment (pastewka) [inline matscipy/sliding_p.py]: documentation missing

File: matscipy/sliding_p.py [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+        h = h1 - h2
+        k = self.C11 * A / h
+        gamma = 2. * np.sqrt(M * k)
+        return M, gamma
+
+
+class SlideWithNormalPressureCuboidCell(object):
+    """ASE constraint used for sliding with pressure coupling
+
+    Only works with diagonal cuboid cells so far.
```
Comment (pastewka) [inline matscipy/sliding_p.py]: Please put an assert or raise an error if those conditions are violated.

File: matscipy/sliding_p.py [hunk 15]
```diff
@@ -0,0 +1,299 @@
...
+    Sliding only works along cell vectors so far.
+    Following L. Pastewka, S. Moser, and M. Moseler,
+    Tribol. Lett. 39, 49 (2010).
+    """
+
+    def __init__(self, top_mask, bottom_mask, Pdir, P, vdir, v, damping):
+        """Constructor.
+
+        top_mask -- boolean numpy array a with a[i] == True for i the index
```
Comment (pastewka) [inline matscipy/sliding_p.py]: docstrings should be numpydoc formatted. See here: https://numpydoc.readthedocs.io/en/latest/format.html

File: matscipy/sliding_p.py [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+        return 0.0
+
+
+class SlideLogger(object):
+    """Logger to be attached to an ASE integrator."""
+
+    def __init__(self, handle, atoms, slider, integrator, step_offset=0):
+        """Constructor.
+
+        handle -- filehandle e.g. pointing to a file opened in w or a mode
```
Comment (pastewka) [inline matscipy/sliding_p.py]: numpydoc

File: matscipy/sliding_p.py [hunk 17]
```diff
@@ -0,0 +1,299 @@
...
+        self.handle = handle
+        self.atoms = atoms
+        self.slider = slider
+        self.integrator = integrator
+        self.step_offset = step_offset
+
+    def write_header(self):
+        self.handle.write('# step | time / fs | T_thermostat / K | P_top / GPa | P_bottom / GPa | h / Ang | v / Ang * fs | a / Ang * fs ** 2 | tau_top / GPa | tau_bottom / GPa\n')
+
+    def __call__(self):
```
Comment (pastewka) [inline matscipy/sliding_p.py]: documentation

File: matscipy/sliding_p.py [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+    tau_top -- shear stress on lid in GPa
+    tau_bottom -- shear stress on base in GPa
+    rows -- all data in a 2d array with axis 0 step and axis 1
+            the values in the order as above
+    """
+
+    def __init__(self, handle):
+        """Constructor.
+
+        handle -- handle or filename pointing to log file
```
Comment (pastewka) [inline matscipy/sliding_p.py]: numpydoc

**Comment by threich** [comment]
Hi Lars,
thanks for the checks, I messed up, sorry for that.

I only intended to merge the barostat (commit ff59a83), I wasn't aware that commits made after the pull request are automatically included in the pull request. The socketcalculator currently only works for python2 and depends on ipi, so I do not think it should be merged into the official matscipy repository at all. I'll create a new pull request only for the barostat.

Sorry once again and thanks for your help.

**Comment by threich** [comment]
Everything related to the socket-calculator should be removed from this MR and the docstring should be numpydoc compliant (at least "pydocstyle --convention=numpy" does not complain anymore) now.

File: matscipy/pressurecoupling.py [hunk 19]
```diff
@@ -396,19 +396,19 @@ class SlideLog(object):
 
     The data of the log files is found as attributes
     (numpy arrays with step as axis):
-    step -- step indices 0, 1, 2, ...
-    time -- simulation time in fs at step
-    T_thermostat -- instantaneous temperature in K from thermostatet region
+    step : step indices 0, 1, 2, ...
```
Comment (pastewka) [inline matscipy/pressurecoupling.py]: Please update to numpydoc, see here: https://numpydoc.readthedocs.io/en/latest/format.html

```
step : int
    Step indices 0, 1, 2, ...
```

The data type follows the colon, not the description. The description is in the next line indented by four spaces. Please update.

Comment (threich) [reply matscipy/pressurecoupling.py]: Thanks for the review, should be ok now.

**Comment by pastewka** [comment]
Thank you for this contribution, @threich
